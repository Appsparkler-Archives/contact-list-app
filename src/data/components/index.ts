export * from "./standard-select-field";
