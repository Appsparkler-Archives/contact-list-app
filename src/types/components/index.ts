export * from "./StandardSelectField";
export * from "./FilterChip";
export * from "./CreateContactForm";
export * from "./ContactListAccordion";
export * from "./ContactForm";
export * from "./DeleteContact";
export * from "./FullScreenModalTransition";
export * from "./ViewContactModal";
