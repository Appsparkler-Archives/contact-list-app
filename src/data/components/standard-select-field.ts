import { IStandSelectFieldOptionItem } from "types/components/StandardSelectField";

export const standardSelectFieldOptions: IStandSelectFieldOptionItem[] = [
  {
    name: "Female",
    value: "female",
  },
  {
    name: "Male",
    value: "male",
  },
];
