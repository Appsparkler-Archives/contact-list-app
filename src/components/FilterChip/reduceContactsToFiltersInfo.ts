import { find, map } from "lodash/fp";
import { reduce } from "lodash/fp";
import { IContactFormData } from "types";

type TContactInfoItemType = "male" | "female" | "business" | "personal";

interface IContactInfoItem {
  type: TContactInfoItemType;
  count: number;
}

export const findContactInfoByType = (type: TContactInfoItemType) =>
  find<IContactInfoItem>((contactInfo) => contactInfo.type === type);

const initialContactInfo: IContactInfoItem[] = [
  {
    type: "female",
    count: 0,
  },
  {
    type: "male",
    count: 0,
  },
  {
    type: "business",
    count: 0,
  },
  {
    type: "personal",
    count: 0,
  },
];

export const reduceContactsToFiltersInfo = reduce<
  IContactFormData,
  IContactInfoItem[]
>((acc, contact) => {
  const mapToUpdateCount = map<IContactInfoItem, IContactInfoItem>(
    (contactInfoItem) => {
      if (contactInfoItem.type === contact.gender) {
        return {
          type: contact.gender,
          count: contactInfoItem.count + 1,
        };
      } else if (contactInfoItem.type === contact.type) {
        return {
          type: contact.type,
          count: contactInfoItem.count + 1,
        };
      } else return contactInfoItem;
    }
  );
  return mapToUpdateCount(acc);
}, initialContactInfo);
