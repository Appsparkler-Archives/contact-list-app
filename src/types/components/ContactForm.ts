import { IContactFormData } from "types";

export type TContactFormFC = React.FC<{
  contact: IContactFormData;
  submitBtnTitle: string;
  formTitle: string;

  // event handlers
  onClickSubmit: (contactData: IContactFormData) => void;
  onClickCancel: () => void;
}>;
