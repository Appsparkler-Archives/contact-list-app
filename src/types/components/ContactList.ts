import React from "react";
import { IContactFormData } from "types";

export type TContactListFC = React.FC<{ contacts: IContactFormData[] }>;
