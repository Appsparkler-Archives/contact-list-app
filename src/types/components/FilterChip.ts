import { IContactFormData } from "types/data";

export type TFiltersFC = React.FC<{
  contacts: IContactFormData[];
}>;
