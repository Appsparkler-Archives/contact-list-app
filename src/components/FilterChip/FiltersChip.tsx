import { Box, Chip } from "@mui/material";
import { useMemo } from "react";
import { TFiltersFC } from "types/components/FilterChip";
import { reduceContactsToFiltersInfo } from "./reduceContactsToFiltersInfo";

export const FiltersChip: TFiltersFC = ({ contacts }) => {
  const contactInfo = useMemo(
    () => reduceContactsToFiltersInfo(contacts),
    [contacts]
  );
  return (
    <Box
      display="flex"
      flexDirection={"row"}
      gap={2}
      justifyContent={"center"}
      p={2}
    >
      {contactInfo.map(({ count, type }) =>
        count > 0 ? (
          <Chip
            key={type}
            label={`${count} ${type}`}
            variant="outlined"
            size="small"
            sx={{ backgroundColor: "background.paper" }}
          />
        ) : null
      )}
    </Box>
  );
};
