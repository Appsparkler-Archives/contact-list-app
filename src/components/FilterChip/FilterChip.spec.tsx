import { Box, Chip } from "@mui/material";
import { useMemo } from "react";
import { TFiltersFC } from "types/components/FilterChip";
import { reduceContactsToFiltersInfo } from "./reduceContactsToFiltersInfo";
import { render, screen } from "@testing-library/react";

const FiltersChip: TFiltersFC = ({ contacts }) => {
  const contactInfo = useMemo(
    () => reduceContactsToFiltersInfo(contacts),
    [contacts]
  );
  return (
    <Box
      display="flex"
      flexDirection={"row"}
      gap={2}
      justifyContent={"center"}
      p={2}
    >
      {contactInfo.map(({ count, type }) =>
        count > 0 ? (
          <Chip
            key={type}
            label={`${count} ${type}`}
            variant="outlined"
            size="small"
          />
        ) : null
      )}
    </Box>
  );
};

describe("Filters - snapshots", () => {
  it("should render nothing except wrapper", () => {
    const { container } = render(<FiltersChip contacts={[]} />);
    expect(container).toMatchSnapshot();
    expect(screen.queryByText("male")).toBeNull();
    expect(screen.queryByText("female")).toBeNull();
    expect(screen.queryByText("business")).toBeNull();
    expect(screen.queryByText("personal")).toBeNull();
  });

  it("should render only 1 female and 1 personal filter chip", () => {
    const { container } = render(
      <FiltersChip
        contacts={[
          {
            gender: "female",
            mobile: "+43295834958345",
            name: "Jenny",
            type: "personal",
          },
        ]}
      />
    );
    expect(container).toMatchSnapshot();
    expect(screen.queryByText("male")).toBeNull();
    expect(screen.getByText("1 female")).toBeTruthy();
    expect(screen.queryByText("business")).toBeNull();
    expect(screen.getByText("1 personal")).toBeTruthy();
  });

  it("should render 2 female, 1 personal and 1 business filter chip", () => {
    const { container } = render(
      <FiltersChip
        contacts={[
          {
            gender: "female",
            mobile: "+43295834958345",
            name: "Jenny",
            type: "personal",
          },
          {
            gender: "female",
            mobile: "+43295834958345",
            name: "Rupa",
            type: "business",
          },
        ]}
      />
    );
    expect(screen.queryByText("male")).toBeNull();
    expect(screen.getByText("2 female")).toBeTruthy();
    expect(screen.getByText("1 business")).toBeTruthy();
    expect(screen.getByText("1 personal")).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
