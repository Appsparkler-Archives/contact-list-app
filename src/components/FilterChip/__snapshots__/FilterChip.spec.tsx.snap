// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters - snapshots should render 2 female, 1 personal and 1 business filter chip 1`] = `
<div>
  <div
    class="MuiBox-root css-hlt1gb"
  >
    <div
      class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-outlinedDefault css-aulpef-MuiChip-root"
    >
      <span
        class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
      >
        2 female
      </span>
    </div>
    <div
      class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-outlinedDefault css-aulpef-MuiChip-root"
    >
      <span
        class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
      >
        1 business
      </span>
    </div>
    <div
      class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-outlinedDefault css-aulpef-MuiChip-root"
    >
      <span
        class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
      >
        1 personal
      </span>
    </div>
  </div>
</div>
`;

exports[`Filters - snapshots should render nothing except wrapper 1`] = `
<div>
  <div
    class="MuiBox-root css-hlt1gb"
  />
</div>
`;

exports[`Filters - snapshots should render only 1 female and 1 personal filter chip 1`] = `
<div>
  <div
    class="MuiBox-root css-hlt1gb"
  >
    <div
      class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-outlinedDefault css-aulpef-MuiChip-root"
    >
      <span
        class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
      >
        1 female
      </span>
    </div>
    <div
      class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-outlinedDefault css-aulpef-MuiChip-root"
    >
      <span
        class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
      >
        1 personal
      </span>
    </div>
  </div>
</div>
`;
